sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala

```diff
@@ -137,6 +137,22 @@ object RowEncoder {
         CreateStruct(convertedFields))
   }
 
+  private def externalDataTypeForInput(dt: DataType): DataType = dt match {
```

Comment:
agree, we should add some doc to say that this method is used to get the external type for a input of Datatype dt, and should only be used in `extractorFor`